Repository: plone.api


Branch: refs/heads/master
Date: 2017-04-05T11:36:01+02:00
Author: Roel Bruggink (jaroel) <roel@jaroel.nl>
Commit: https://github.com/plone/plone.api/commit/6c6fb995896ebce9284b0cac42f177b461c1350a

Add failing test

Files changed:
M src/plone/api/content.py
M src/plone/api/tests/test_content.py

diff --git a/src/plone/api/content.py b/src/plone/api/content.py
index e6f1aa3..196e86a 100644
--- a/src/plone/api/content.py
+++ b/src/plone/api/content.py
@@ -261,7 +261,10 @@ def copy(source=None, target=None, id=None, safe_id=False):
 
     new_id = copy_info[0]['new_id']
     if id:
-        return rename(obj=target[new_id], new_id=id, safe_id=safe_id)
+        if not safe_id and id in target:
+            raise InvalidParameterError
+        else:
+            return rename(obj=target[new_id], new_id=id, safe_id=safe_id)
     else:
         return target[new_id]
 
diff --git a/src/plone/api/tests/test_content.py b/src/plone/api/tests/test_content.py
index da25b89..47bca78 100644
--- a/src/plone/api/tests/test_content.py
+++ b/src/plone/api/tests/test_content.py
@@ -594,6 +594,17 @@ def test_copy(self):
         assert container['bargains']['item'].aq_base is bargain.aq_base
         assert container['products']['item']
 
+    def test_copy_same_id(self):
+        obj = self.contact
+
+        # Using the same id should fail
+        from plone.api.exc import InvalidParameterError
+        with self.assertRaises(InvalidParameterError):
+            api.content.copy(obj, obj.__parent__, obj.id)
+
+        # Using safe_id=True should work
+        api.content.copy(obj, obj.__parent__, obj.id, safe_id=True)
+
     def test_delete_constraints(self):
         """Test the constraints for deleting content."""
 


Repository: plone.api


Branch: refs/heads/master
Date: 2017-04-05T11:36:27+02:00
Author: Roel Bruggink (jaroel) <roel@jaroel.nl>
Commit: https://github.com/plone/plone.api/commit/bed9aa61fd04d59208564f02d361c61394fa6022

changelog

Files changed:
M CHANGES.rst

diff --git a/CHANGES.rst b/CHANGES.rst
index eb7e481..67e0619 100644
--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -17,6 +17,9 @@ Bug fixes:
 - Simplify the ``plone.api.content.delete`` method.
   [thet]
 
+- content.copy with safe_id=False should raise it's own exeception. Fixes #340
+  [jaroel]
+
 
 1.6.1 (2017-03-31)
 ------------------


Repository: plone.api


Branch: refs/heads/master
Date: 2017-04-05T14:37:17+02:00
Author: Roel Bruggink (jaroel) <roel@jaroel.nl>
Commit: https://github.com/plone/plone.api/commit/c4db1cd1b9b44ec7636c48c48c3f326ab0dce3fa

Add error message

Files changed:
M src/plone/api/content.py

diff --git a/src/plone/api/content.py b/src/plone/api/content.py
index 196e86a..f219a80 100644
--- a/src/plone/api/content.py
+++ b/src/plone/api/content.py
@@ -262,7 +262,8 @@ def copy(source=None, target=None, id=None, safe_id=False):
     new_id = copy_info[0]['new_id']
     if id:
         if not safe_id and id in target:
-            raise InvalidParameterError
+            msg = "Duplicate ID '{0}' in '{1}' for '{2}'"
+            raise InvalidParameterError(msg.format(id, target, source))
         else:
             return rename(obj=target[new_id], new_id=id, safe_id=safe_id)
     else:


Repository: plone.api


Branch: refs/heads/master
Date: 2017-04-06T18:09:51+02:00
Author: Jens W. Klein (jensens) <jk@kleinundpartner.at>
Commit: https://github.com/plone/plone.api/commit/8d249811cf8d56eb89c0dce0dcd782d72b05a4d0

Merge pull request #365 from plone/340-copy_safe_id

Raise InvalidParameterError on duplicate id

Files changed:
M CHANGES.rst
M src/plone/api/content.py
M src/plone/api/tests/test_content.py

diff --git a/CHANGES.rst b/CHANGES.rst
index eb7e481..67e0619 100644
--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -17,6 +17,9 @@ Bug fixes:
 - Simplify the ``plone.api.content.delete`` method.
   [thet]
 
+- content.copy with safe_id=False should raise it's own exeception. Fixes #340
+  [jaroel]
+
 
 1.6.1 (2017-03-31)
 ------------------
diff --git a/src/plone/api/content.py b/src/plone/api/content.py
index e6f1aa3..f219a80 100644
--- a/src/plone/api/content.py
+++ b/src/plone/api/content.py
@@ -261,7 +261,11 @@ def copy(source=None, target=None, id=None, safe_id=False):
 
     new_id = copy_info[0]['new_id']
     if id:
-        return rename(obj=target[new_id], new_id=id, safe_id=safe_id)
+        if not safe_id and id in target:
+            msg = "Duplicate ID '{0}' in '{1}' for '{2}'"
+            raise InvalidParameterError(msg.format(id, target, source))
+        else:
+            return rename(obj=target[new_id], new_id=id, safe_id=safe_id)
     else:
         return target[new_id]
 
diff --git a/src/plone/api/tests/test_content.py b/src/plone/api/tests/test_content.py
index da25b89..47bca78 100644
--- a/src/plone/api/tests/test_content.py
+++ b/src/plone/api/tests/test_content.py
@@ -594,6 +594,17 @@ def test_copy(self):
         assert container['bargains']['item'].aq_base is bargain.aq_base
         assert container['products']['item']
 
+    def test_copy_same_id(self):
+        obj = self.contact
+
+        # Using the same id should fail
+        from plone.api.exc import InvalidParameterError
+        with self.assertRaises(InvalidParameterError):
+            api.content.copy(obj, obj.__parent__, obj.id)
+
+        # Using safe_id=True should work
+        api.content.copy(obj, obj.__parent__, obj.id, safe_id=True)
+
     def test_delete_constraints(self):
         """Test the constraints for deleting content."""
 


